crates/ibc/src/core/ics03_connection/handler/conn_open_confirm.rs

```diff
@@ -1,15 +1,174 @@
 //! Protocol logic specific to processing ICS3 messages of type `MsgConnectionOpenConfirm`.
 
+use crate::core::context::ContextError;
 use crate::core::ics03_connection::connection::{ConnectionEnd, Counterparty, State};
 use crate::core::ics03_connection::context::ConnectionReader;
 use crate::core::ics03_connection::error::ConnectionError;
 use crate::core::ics03_connection::events::OpenConfirm;
 use crate::core::ics03_connection::handler::{ConnectionIdState, ConnectionResult};
 use crate::core::ics03_connection::msgs::conn_open_confirm::MsgConnectionOpenConfirm;
+use crate::core::ics24_host::identifier::{ClientId, ConnectionId};
+use crate::core::ics24_host::path::ConnectionsPath;
+use crate::core::{ExecutionContext, ValidationContext};
 use crate::events::IbcEvent;
 use crate::handler::{HandlerOutput, HandlerResult};
 use crate::prelude::*;
 
+pub(crate) fn validate<Ctx>(ctx_b: &Ctx, msg: &MsgConnectionOpenConfirm) -> Result<(), ContextError>
+where
+    Ctx: ValidationContext,
+{
+    let vars = LocalVars::new(ctx_b, msg)?;
+    validate_impl(ctx_b, msg, &vars)
+}
+
+fn validate_impl<Ctx>(
+    ctx_b: &Ctx,
+    msg: &MsgConnectionOpenConfirm,
+    vars: &LocalVars,
+) -> Result<(), ContextError>
+where
+    Ctx: ValidationContext,
+{
+    let conn_end_on_b = vars.conn_end_on_b();
+    if !conn_end_on_b.state_matches(&State::TryOpen) {
+        return Err(ContextError::ConnectionError(
+            ConnectionError::ConnectionMismatch {
+                connection_id: msg.conn_id_on_b.clone(),
+            },
+        ));
```

Comment:
I think it'll be nice to use the `Into` impl for converting `ConnectionError` to `ContextError` (here and elsewhere) ->
```suggestion
        return Err(ConnectionError::InvalidConsensusHeight {
            target_height: msg.consensus_height_of_a_on_b,
            current_height: host_height,
        }
        .into());
```

Reply:
Yes much cleaner [093cc9e](https://github.com/cosmos/ibc-rs/pull/257/commits/093cc9efc5f0869fca8ea3b3ee9492819aa9e395)